### PR TITLE
Support es256

### LINKED ISSUE #941: Support ES256 algorithm
Google IAP uses ES256;
https://cloud.google.com/iap/docs/signed-headers-howto

Looks like currently only RS256 support is provided.

### PR DESCRIPTION
**Pull request by JimmyCYJ**
**What this PR does / why we need it**:Add support for ES256

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #941 

**Special notes for your reviewer**:
Add support for ES256 algorithm.
Add check for "kty" parameters.
Add basic unit tests.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: src/envoy/http/jwt_auth/jwt.cc [hunk 1]
```diff
@@ -421,27 +485,67 @@ void Pubkeys::CreateFromJwksCore(const std::string &pkey_jwks) {
...
+    ExtractPubkeyFromJwkEC(jwk_json);
+  } else if (kty == "RSA" && alg.compare(0, 2, "RS") == 0) {
+    ExtractPubkeyFromJwkRSA(jwk_json);
+  }
+}
+
+void Pubkeys::ExtractPubkeyFromJwkRSA(Json::ObjectSharedPtr jwk_json) {
+  std::unique_ptr<Pubkey> pubkey(new Pubkey());
+  pubkey->kid_ = jwk_json->getString("kid");
+  pubkey->alg_ = jwk_json->getString("alg");
```
Comment (qiwzhang) [inline src/envoy/http/jwt_auth/jwt.cc]: need to try and catch to handle invalid format

File: src/envoy/http/jwt_auth/jwt_test.cc [hunk 2]
```diff
@@ -547,6 +619,20 @@ TEST_F(JwtTestJwks, JwkBadPublicKey) {
...
+  DoTest(ds.kTokenEC, ds.kPublicKeyJwkEC, "jwks", true, Status::OK, payload);
+  // ES256-signed token without kid specified.
+  DoTest(ds.kTokenECNoKid, ds.kPublicKeyJwkEC, "jwks", true, Status::OK,
+         payload);
+}
+
+TEST_F(JwtTestJwks, NonExistKidEC) {
+  DoTest(ds.kJwtWithNonExistKidEC, ds.kPublicKeyJwkEC, "jwks", false,
+         Status::KID_ALG_UNMATCH, nullptr);
+}
```
Comment (qiwzhang) [inline src/envoy/http/jwt_auth/jwt_test.cc]: please add some negative cases 1) pubkey wrong formats. 2) token wrong format

Comment (JimmyCYJ) [reply src/envoy/http/jwt_auth/jwt_test.cc]: Tests are added. Thanks.

File: src/envoy/http/jwt_auth/jwt.cc (hunk 1 above)
Comment (JimmyCYJ) [reply src/envoy/http/jwt_auth/jwt.cc]: Done. Thanks.

**Comment by JimmyCYJ** [comment]
Please take a look. Thanks!

File: src/envoy/http/jwt_auth/jwt.h [hunk 3]
```diff
@@ -31,74 +32,83 @@ enum class Status {
   OK = 0,
 
   // JWT token is required.
-  JWT_MISSED,
+  JWT_MISSED = 1,
```
Comment (qiwzhang) [inline src/envoy/http/jwt_auth/jwt.h]: not need to add =1,  
let compiler assign them.

In my PR,  I will remove BEARER_PREFIX_MISMATCH,  then I would have to change each assignment  too

Comment (JimmyCYJ) [reply src/envoy/http/jwt_auth/jwt.h]: Per offline discussion, I use assignment here to make it easier for debugging. This enum class is updated.

**Pull request merged**